app/scripts/modules/core/src/presentation/formsy/components/ExpressionInput.tsx

```diff
@@ -0,0 +1,160 @@
+import { truncate } from 'lodash';
+
+import { FormsyComponent, IFormsyComponentProps, IFormComponentState } from '../FormsyComponent';
+import { Markdown } from '../../Markdown';
+
+import '../spel2js.templateParser';
+import { Modal } from 'react-bootstrap';
+
+export interface IExpressionInputProps extends IFormsyComponentProps {
+  placeholder?: string;
+  context?: object;
+  locals?: object;
+  markdown?: boolean;
+}
+
+export interface IExpressionInputState extends IFormComponentState {
+  showContextModal: boolean;
+  spelPreview: string;
+  spelError: {
+    message: string;
+    context: string;
+    contextTruncated: string;
+  };
+}
+
+/**
+ * A validating Formsy form component for SpEL Expressions
+ */
+@autoBindMethods
+export class ExpressionInput extends FormsyComponent<string, IExpressionInputProps, IExpressionInputState> {
+  public static contextTypes = FormsyComponent.contextTypes;
+  public static defaultProps = Object.assign({ context: {}, locals: {} }, FormsyComponent.defaultProps);
+
+  public validate(props: IExpressionInputProps = this.props): boolean {
+    const value = this.getValue();
+    const { context, locals } = props;
+
+    const stringify = (obj: any): string => {
+      return obj === null ? 'null' :
+        obj === undefined ? 'undefined' :
+          JSON.stringify(obj, null, 2);
+    };
+
+    try {
+      if (!value) {
+        return true;
+      }
+
+      const exprs = spel2js.TemplateParser.parse(value);
+      const results = exprs.map(expr => expr.eval(context, locals));
+      this.setState({ spelError: null, spelPreview: results.join('') });
+
+      return true;
+    } catch (err) {
+      const spelError = {
+        message: null as string,
+        context: null as string,
+        contextTruncated: null as string,
+      };
+
+      if (err.name && err.message) {
+        if (err.name === 'NullPointerException' && err.state && err.state.activeContext) {
+          spelError.context = stringify(err.state.activeContext.peek());
+          spelError.contextTruncated = truncate(spelError.context, { length: 200 });
+        }
+        spelError.message = `${err.name}: ${err.message}`;
+      } else {
+        try {
+          spelError.message = JSON.stringify(err);
+        } catch (ignored) {
+          spelError.message = err.toString();
+        }
+      }
+
+      this.setState({ spelError, spelPreview: null });
+      return false;
+    }
+  }
+
+  public componentWillReceiveProps(nextProps: IExpressionInputProps): void {
+    this.validate(nextProps);
+  }
+
+  public getErrorMessage(): string {
+    const { spelError } = this.state;
+    if (spelError) {
+      return spelError.message + (spelError.contextTruncated ? ' -- ' + spelError.contextTruncated : '');
+    }
+
+    return super.getErrorMessage();
+  }
+
+  private hideContextModal(): void {
+    this.setState({ showContextModal: false })
+  }
+
+  private showSpelErrorModal(): void {
```

Comment:
show and hide functions have very different names but toggle the same state property; maybe name them similarly?